### PR TITLE
[DONT MERGE] Testing approval-distribution concurrent workers.

### PR DESCRIPTION
**Pull request by sandreim**

### CODE REVIEW

File: node/network/approval-distribution/src/lib.rs [hunk 1]
```diff
@@ -1674,83 +1811,78 @@ async fn adjust_required_routing_and_propagate(
...
 	}
 
-	async fn run<Context>(self, ctx: Context)
-	where
-		Context: SubsystemContext<Message = ApprovalDistributionMessage>,
-		Context: overseer::SubsystemContext<Message = ApprovalDistributionMessage>,
-	{
-		let mut state = State::default();
-
+	#[tokio::main]
```
Comment (ordian) [inline node/network/approval-distribution/src/lib.rs:1857]: what's the purpose of tokio::main here considering we're manually spawning the threads? (I understand that both things are not what we want to do in production and are only used for this test, but curious why ::main is needed at all)

Comment (sandreim) [reply node/network/approval-distribution/src/lib.rs:1857]: It is not needed, it is just a leftover. I'll switch it to tasks. I hoped that I can hold the lock while awaiting if I use threads...

File: node/network/approval-distribution/src/lib.rs [hunk 2]
```diff
@@ -233,31 +227,31 @@ enum Resend {
 ///
 /// It tracks metadata about our view of the unfinalized chain,
 /// which assignments and approvals we have seen, and our peers' views.
-#[derive(Default)]
+#[derive(Clone, Default)]
 struct State {
 	/// These two fields are used in conjunction to construct a view over the unfinalized chain.
-	blocks_by_number: BTreeMap<BlockNumber, Vec<Hash>>,
-	blocks: HashMap<Hash, BlockEntry>,
+	blocks_by_number: Arc<Mutex<BTreeMap<BlockNumber, Vec<Hash>>>>,
```
Comment (rphmeier) [inline node/network/approval-distribution/src/lib.rs:233]: This many mutexes scares me, frankly. Lock order needs to be well-defined to avoid deadlocks.

Comment (rphmeier) [reply node/network/approval-distribution/src/lib.rs:233]: We should just `Arc` the whole state, not `Arc` everything in the state.

Comment (sandreim) [reply node/network/approval-distribution/src/lib.rs:233]: You are right. I wanted to go for more granular locking as I did not know that we can't await with holding the mutex, but now that this is no longer happening, I can just lock the whole state.

Comment (rphmeier) [reply node/network/approval-distribution/src/lib.rs:233]: That's actually not what I meant.

```rust
state: Arc<State>

struct State {
    a: Mutex<A>,
    b: Mutex<B>,
}
```

vs 

```rust
state: State

struct State {
    a: Arc<Mutex<A>>,
    b: Arc<Mutex<B>>,
}
```

vs 

```rust
state: Arc<Mutex<State>

struct State {
    a: A,
    b: B,
}
```

are all different.

Comment (sandreim) [reply node/network/approval-distribution/src/lib.rs:233]: Yup 🤯

**Pull request closed without merging**